analytics/location_heatmaps/geo_utils.py

```diff
@@ -217,43 +288,129 @@ def split_regions(tree_prefix_list,
-        if not new_tree.has_key(new_prefix):
-          fresh_expand += 1
-          new_tree[new_prefix] = len(new_tree_prefix_list)
-          new_tree_prefix_list.append(new_prefix)
-    else:
-      if collapse_threshold is not None and\
-          count <= collapse_threshold and\
-          len(prefix) > 2:
-
-        old_prefix = prefix[:-3]
-        collapsed += 1
-        if not new_tree.has_key(old_prefix):
-          created += 1
-          new_tree[old_prefix] = len(new_tree_prefix_list)
-          new_tree_prefix_list.append(old_prefix)
+  if positivity:
+    for i in range(0, len(tree_prefix_list), 2):
+      if expand_all:
+        neg_count = threshold + 1
+        pos_count = threshold + 1
+      else:
+        neg_count = vector_counts[i]
+        pos_count = vector_counts[i + 1]
+      neg_prefix = tree_prefix_list[i]
+      pos_prefix = tree_prefix_list[i + 1]
+
+      # check whether the tree has reached the bottom
+      if len(pos_prefix.split('/')) >= image_bit_level:
+        continue
+
+      # total = pos_count + neg_count
+      # p = pos_count / total
+      # confidence = np.sqrt((1-p)*p/total)
+      # error bound propagation.
+      # confidence +/- noise
+      # pos_count/total +/- (confidence+conf_noise) => 95% interval for 95% noise interval.
+
+      if pos_count > threshold and neg_count > threshold:
+        neg_child = get_default_children(positivity, split='neg')
+        pos_child = get_default_children(positivity, split='pos')
+        for j in range(len(pos_child)):
+          new_prefix = f'{neg_prefix}/{neg_child[j]}'
+          if not new_tree.has_key(new_prefix):
+            fresh_expand += 1
+            new_tree[new_prefix] = len(new_tree_prefix_list)
+            new_tree_prefix_list.append(new_prefix)
+
+            new_prefix = f'{pos_prefix}/{pos_child[j]}'
+            new_tree[new_prefix] = len(new_tree_prefix_list)
+            new_tree_prefix_list.append(new_prefix)
+      else:
+        if collapse_threshold is not None and \
+            (
+                pos_count < collapse_threshold or neg_count < collapse_threshold) and \
+            len(pos_prefix) > 3 and len(neg_prefix) > 3:
+
+          old_prefix = neg_prefix[:-4]
+          collapsed += 1
+          if not new_tree.has_key(old_prefix):
+            created += 1
+            new_tree[old_prefix] = len(new_tree_prefix_list)
+            new_tree_prefix_list.append(old_prefix)
+
+            old_prefix = pos_prefix[:-4]
+            new_tree[old_prefix] = len(new_tree_prefix_list)
+            new_tree_prefix_list.append(old_prefix)
+        else:
+          unchanged += 1
+          new_tree[f'{neg_prefix}'] = len(new_tree_prefix_list)
+          new_tree_prefix_list.append(f'{neg_prefix}')
+          new_tree[f'{pos_prefix}'] = len(new_tree_prefix_list)
+          new_tree_prefix_list.append(f'{pos_prefix}')
+  else:
+    for i in range(len(tree_prefix_list)):
+      if expand_all:
+        count = threshold + 1
```

Comment:
nit: I'd add an explanatory comment here along the lines of "just pretend we have enough reports in this node to expand so that we can easily reuse code below"